### PR TITLE
NEW Adding a shuffle method to ArrayList

### PR DESCRIPTION
**Pull request by ScopeyNZ**
I considered adding some `applyCallback(callable $callable)` API but I couldn't think of any other use case for it. It's different from `each` because we want one callback on the whole list... not each individual item.

### CODE REVIEW

File: src/ORM/ArrayList.php [hunk 1]
```diff
@@ -572,6 +572,18 @@ public function sort()
...
+    /**
+     * Shuffle the items in this array list
+     *
+     * @return $this
+     */
+    public function shuffle()
+    {
+        shuffle($this->items);
+
+        return $this;
```
Comment (robbieaverill) [inline src/ORM/ArrayList.php]: Some of the other methods in this class are immutable, do you want this to be as well?

**Comment by robbieaverill** [comment]
cc @silverstripe/core-team as an API change

**Comment by dnsl48** [comment]
I thought that's not a change, but new API method though?

**Comment by robbieaverill** [comment]
Yeah, sorry - I'm not saying it's a breaking change, but it is still a change to the API surface

File: src/ORM/ArrayList.php (hunk 1 above)
Comment (ScopeyNZ) [reply src/ORM/ArrayList.php]: Given the nature of PHP's `shuffle` - I don't really think it should be. Happy to adjust if asked.

Comment (dnsl48) [reply src/ORM/ArrayList.php]: That's the whole point of this function - to reorder in place, without allocating extra memory. Otherwise we could do 
```php
$array = $arrayList->toArray();
shuffle($array);
return new ArrayList($array);
```

File: tests/php/ORM/ArrayListTest.php [hunk 2]
```diff
@@ -1236,4 +1236,15 @@ public function testDataClass()
...
+
+    public function testShuffle()
+    {
+        $list = new ArrayList(range(1, 10));
+
+        $list->shuffle();
+
+        for ($i = 1; $i <= 10; $i++) {
+            $this->assertContains($i, $list);
+        }
```
Comment (unclecheese) [inline tests/php/ORM/ArrayListTest.php]: The odds of this coming back in the same order are 1 / 3,600,000. Should we have it test the expected result (a shuffle) and allow 1 in 3.6M tests to fail curiously? If not, set the range to 50 and if the probability gods rain down up on us, we deserve it.

Comment (dnsl48) [reply tests/php/ORM/ArrayListTest.php]: but there's no criterion saying it must not repeat the original list. that can be considered as just 1 more possible combination

Comment (ScopeyNZ) [reply tests/php/ORM/ArrayListTest.php]: Yeah I didn't bother testing that it actually shuffles - I feel like we're just testing PHP functionality at that point. If there was some way I could assert that the `shuffle` line is being called though... 

I'm happy to up the range (a lot) and add a `$this->assertNotEquals()` or whatever I need to do for array comparison 😧

Comment (dnsl48) [reply tests/php/ORM/ArrayListTest.php]: simply saying that ideally tests should reflect the functionality requirements "as is".
Otherwise, if we add such test, we may need to add that feature to the function implementation (e.g. calculate a hash of the list and after shuffling check the hash of the new list is different, otherwise repeat the shuffle).

Comment (unclecheese) [reply tests/php/ORM/ArrayListTest.php]: It's just weird that the test doesn't test what it's supposed to do. If I went in and changed the `shuffle()` method to `return $this;`, the tests would pass.

Comment (dnsl48) [reply tests/php/ORM/ArrayListTest.php]: agree. what if we test that 2 shuffles are not the same? :) I guess that's still out of our control (as RNG is hidden under php), but at least we don't exclude the `same as the original` variant from the possible outcomes?

Comment (unclecheese) [reply tests/php/ORM/ArrayListTest.php]: If we changed the range to 50, we would be approaching UUID entropies. A false negative is never, ever, ever, ever going to happen.

**Comment by sminnee** [comment]
>  I think you should label this as an API change as well

I thought that was just for API breakages?

**Comment by sminnee** [comment]
Is it worth adding DataList::shuffle() that calls `$this->sort('RAND()')`, as well, so the two APIs have parity?

**Comment by unclecheese** [comment]
@sminnee That was my thought, too, but is `RAND()` cross platform?

**Comment by ScopeyNZ** [comment]
Annoyingly it seems to be `random()` in PostgreSQL but `rand()` in MySQL.

**Comment by unclecheese** [comment]
That's what I remember. Worth thinking about down the track, and maybe putting into `SS_List`.

**Comment by dnsl48** [comment]
Unfortunately, that's not ANSI SQL - https://stackoverflow.com/a/1209946

File: tests/php/ORM/ArrayListTest.php (hunk 2 above)
Comment (dnsl48) [reply tests/php/ORM/ArrayListTest.php]: I'm not that worried about false-positives, but rather thinking that tests should describe functional requirements. Currently, if we add a simple check that shuffled != original, then it won't be reflecting the truth, as we don't have this functionality implemented.

I'm only saying that, because I already had the problem in understanding whether test actually reflects requirements, or simply trying to check some side-effects. In particular this one - https://github.com/silverstripe/silverstripe-framework/blob/4/tests/php/ORM/DataListTest.php#L614

After a while, people maintaining the code base won't know whether we actually have this functional requirement, or do we just test it for having test coverage.

Comment (unclecheese) [reply tests/php/ORM/ArrayListTest.php]:
```php
$shuffle1 = $list->shuffle();
$shuffle2 = $list->shuffle();

$shuffle1 !== $shuffle2
```

Does that not accurately describe a shuffle, barring one extremely, out-of-this-universe edge case?

Comment (dnsl48) [reply tests/php/ORM/ArrayListTest.php]: haha, sorry for being a pain... but yes, this appears to be what I imagine :)

**Comment by sminnee** [comment]
> Unfortunately, that's not ANSI SQL

`$this->sort(DB::getConn()->random())` then. ;-)

https://github.com/silverstripe/silverstripe-framework/blob/4/src/ORM/Connect/MySQLDatabase.php#L522
https://github.com/silverstripe/silverstripe-postgresql/blob/2/code/PostgreSQLDatabase.php#L708
https://github.com/silverstripe/silverstripe-sqlite3/blob/2/code/SQLite3Database.php#L276

> maybe putting into SS_List.

That breaks any other SS_List implementation, so not in 4.

File: tests/php/ORM/ArrayListTest.php (hunk 2 above)
Comment (ScopeyNZ) [reply tests/php/ORM/ArrayListTest.php]: Ok I've updated the test. Asserting two shuffles are different isn't really any different from asserting a shuffle isn't the same as the original order - I've also up-ed the range to 50.

**Comment by unclecheese** [comment]
Yeah, I meant in 5.0 :)

**Comment by robbieaverill** [comment]
@sminnee 

>I thought that was just for API breakages?



Nah, it categorises changes under "API Changes" in changelogs. See [the docs](https://docs.silverstripe.org/en/4/contributing/code/#commit-messages).

**Comment by ScopeyNZ** [comment]
I've added an update to DataList too now

**Comment by unclecheese** [comment]
If the test doesn't pass, I'm buying a lotto ticket.
